### ISSUE TITLE
Behaviour with ECMA private fields? 

### ISSUE THREAD

**Issue by seivan** [mobxjs/mobx#3573]
This is a follow up on https://github.com/mobxjs/mobx/issues/2339
but to summarize, with the introduction of the `accessor` keyword introduced in TS 4.9, how will Mobx behave with ECMA private fields (`#__someThingPrivate`)? 

Since the `accessor` keyword automatically de-sugars into getters and setters, my take would be that the backing field isn't an observable, but since it has a getter and a setter, it will create computed and actions, is this correct?

**Comment by mweststrate** [comment]
MobX relies upon reflection, and private fields don't show up in reflection
afaik, so nothing at all would happen with them and they'd be
non-observable / lost based on the context in which you are using them

**Comment by seivan** [comment]
> MobX relies upon reflection, and private fields don't show up in reflection afaik, so nothing at all would happen with them and they'd be non-observable / lost based on the context in which you are using them
> […](#)

Right, but what about the generated getters and setters? Wouldn't the getters at least count as a computed? 

If not, with the introduction of the new keyword which (unfortunately is using ECMA private fields) is it possible to have getter & setters to automatically be treated as computed/observables & actions?

**Comment by mweststrate** [comment]
> is it possible to have getter & setters to automatically be treated as computed/observables & actions?

You can already use `false` or `computed` to explicitly specify what MobX should do with a field. 

> , with the introduction of the accessor keyword introduced in TS 4.9

I didn't see anything in the changelog, could you link what you are referring to?

**Comment by seivan** [comment]
> > is it possible to have getter & setters to automatically be treated as computed/observables & actions?
> 
> You can already use `false` or `computed` to explicitly specify what MobX should do with a field.

Well was hoping it be automatic, given how getters and setters are automatic as of today, the only difference is that the backing value isn't observable. Which is a shame. 
 
> > , with the introduction of the accessor keyword introduced in TS 4.9
> 
> I didn't see anything in the changelog, could you link what you are referring to?

https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-rc/#auto-accessors-in-classes

Sorry for the delay!

**Comment by mweststrate** [comment]
I'm a bit rusty on the new decorators proposal (worked on it but it is a while ago), we might be able to intercept this with decorators, but least not with `makeAutoObservable` and friends, as we won't be able to either make the private observable, or even detect that this is an accessor, not an ordinary getter

**Comment by seivan** [comment]
Isn’t makeAutoObservable runtime? I’ve been meaning to ask, if you have a regular property that’s not an observable and makeAuto is configured to ignore it following by have a getter and a setter - will they be mobxified? Like will the getter behave like a computed and will the setter be wrapped in an action? Because if that happens and makeAuto just loads up getters/setters at runtime (when sugared) then shouldn’t it work out of the box? Sent from an iPhone on the go. On 15 Nov 2022, at 09:24, Michel Weststrate ***@***.***> wrote:﻿
I'm a bit rusty on the new decorators proposal (worked on it but it is a while ago), we might be able to intercept this with decorators, but least not with makeAutoObservable and friends, as we won't be able to either make the private observable, or even detect that this is an accessor, not an ordinary getter

**Comment by mweststrate** [comment]
Eh yeah I think you are right, we reuse the closure so that should work :)

**Comment by mweststrate** [comment]
It will just be meaningless as the backing field is not observable, so the
computed / action doesn't buy you any reactivity

**Comment by justinfagnani** [comment]
fwiw, standard decorators work fine with private auto-accessors. The decorator receives an `access` object that has get() and set() methods to the backing store.

**Comment by kubk** [comment]
Modern decorators support ECMA private fields, so `@observable accessor #field` works. Related test: https://github.com/mobxjs/mobx/pull/4668

There's no way to support them in `makeObservable` / `makeAutoObservable`. As a workaround, in TS we can use `private` fields instead and pass the extra generic when needed, for example:

```ts
makeAutoObservable<this, "privateTsField">(this, { privateTsField: false })
```
